kernel/posix/pthread.c

```diff
@@ -0,0 +1,550 @@
+}
+
+/**
+ * @brief Create a new thread.
+ *
+ * See IEEE 1003.1
+ */
+int pthread_create(pthread_t *newthread, const pthread_attr_t *attr,
+		   void *(*threadroutine)(void *), void *arg)
+{
+	static atomic_val_t thread_count;
+	atomic_val_t thread_index = 0;
+	s32_t prio;
+	pthread_mutexattr_t att;
+	pthread_condattr_t cond_attr;
+	struct posix_thread *thread;
+
+	if (!attr || !attr->initialized || !attr->stack || !attr->stacksize) {
+		errno = EINVAL;
+		return -1;
+	}
+
+	prio = posix_to_zephyr_priority(attr->priority, attr->schedpolicy);
+	thread_index = atomic_inc(&thread_count);
+
+	/* FIXME: Thread space allocation need to use mempool for better
+	 * space utilization under limited number of threads.
+	 */
+	if (thread_index < CONFIG_MAX_PTHREAD_COUNT) {
+		thread = &pthread[thread_index];
+		thread->cancel_state = (1 << PTHREAD_CANCEL_POS) & attr->flags;
+		thread->state = attr->detachstate;
+		thread->cancel_pending = 0;
+		thread->state_lock.sem = &thread->state_lock_sem;
+		pthread_mutex_init(&thread->state_lock, &att);
+		thread->cancel_lock.sem = &thread->cancel_lock_sem;
+		pthread_mutex_init(&thread->cancel_lock, &att);
+		pthread_cond_init(&thread->state_cond, &cond_attr);
+
+		*newthread = (pthread_t)
+				k_thread_create(&thread->thread,
+						attr->stack, attr->stacksize,
+						(k_thread_entry_t)
+						zephyr_thread_wrapper,
+						(void *)arg, NULL,
+						threadroutine, prio,
+						attr->flags,
+						attr->delayedstart);
+
+	} else {
+		/* Insuffecient resource, maximum thread count reached */
+		errno = EAGAIN;
+		return -1;
+	}
+
+	return 0;
+}
+
+
+/**
+ * @brief Set cancelability State.
+ *
+ * See IEEE 1003.1
+ */
+int pthread_setcancelstate(int state, int *oldstate)
+{
+	struct posix_thread *pthread = (struct posix_thread *) pthread_self();
+
+	if (state != PTHREAD_CANCEL_ENABLE &&
+	    state != PTHREAD_CANCEL_DISABLE) {
+		errno = EINVAL;
+		return -1;
+	}
+
+	*oldstate = pthread->cancel_state;
+	pthread_mutex_lock(&pthread->cancel_lock);
+	pthread->cancel_state = state;
+
+	if (state == PTHREAD_CANCEL_DISABLE) {
+		_thread_essential_set();
+	} else {
+		_thread_essential_clear();
+	}
+
+	pthread_mutex_unlock(&pthread->cancel_lock);
+
+	if (state == PTHREAD_CANCEL_ENABLE && pthread->cancel_pending) {
+		pthread_exit((void *)PTHREAD_CANCELED);
+	}
+
+	return 0;
+}
+
+/**
+ * @brief Cancel execution of a thread.
+ *
+ * See IEEE 1003.1
+ */
+int pthread_cancel(pthread_t pthread)
+{
+	struct posix_thread *thread = (struct posix_thread *) pthread;
+
+	if (thread == NULL || thread->state == PTHREAD_TERMINATED) {
+		errno = ESRCH;
+		return -1;
+	}
+
+	pthread_mutex_lock(&thread->cancel_lock);
+	thread->cancel_pending = 1;
+
+	if (thread->cancel_state == PTHREAD_CANCEL_ENABLE) {
+		if (thread->state == PTHREAD_DETACHED) {
+			thread->state = PTHREAD_TERMINATED;
+			pthread_mutex_unlock(&thread->cancel_lock);
+			k_thread_abort((k_tid_t) thread);
+		} else {
+			thread->retval = PTHREAD_CANCELED;
```

Comment:
Shouldn't this be thread->state?

Reply:
Yes I updated this - 
If a thread with CANCEL_ENABLE and JOINABLE is cancelled, It broadcast the join condition (for threads waiting to join it) and change the state to PTHREAD_EXIT. (something like thread_exit()), Please give feedback if this is ok we need to make some changes in this ?